contracts/PvpArena.sol

```diff
@@ -728,124 +688,214 @@ contract PvpArena is Initializable, AccessControlUpgradeable {
-    /// @dev wether or not the character is still in time to start a duel
-    function isCharacterWithinDecisionTime(uint256 characterID)
-        public
-        view
-        returns (bool)
-    {
-        return
-            duelByAttacker[characterID].createdAt.add(decisionSeconds) >
-            block.timestamp;
+    /// @dev returns ranked prize percentages distribution
+    function getPrizePercentages() external view returns (uint256[] memory) {
+        return prizePercentages;
     }
 
-    /// @dev wether or not the character is the attacker in a duel
-    // and has not performed an action
-    function hasPendingDuel(uint256 characterID) public view returns (bool) {
-        return duelByAttacker[characterID].isPending;
+    /// @dev returns the account's ranking prize pool earnings
+    function getPlayerPrizePoolRewards() public view returns (uint256) {
+        return _rankingRewardsByPlayer[msg.sender];
     }
 
-    /// @dev wether or not a character can appear as someone's opponent
-    function isCharacterAttackable(uint256 characterID)
-        public
-        view
-        returns (bool)
-    {
-        uint256 lastActivity = _lastActivityByCharacter[characterID];
+    /// @dev returns the current duel queue
+    function getDuelQueue() public view returns (uint256[] memory) {
+        uint256 length = _duelQueue.length();
+        uint256[] memory values = new uint256[](length);
 
-        return lastActivity.add(unattackableSeconds) <= block.timestamp && !_duelQueue.contains(characterID) && !characterDefending[characterID];    
+        for (uint256 i = 0; i < length; i++) {
+            values[i] = _duelQueue.at(i);
+        }
+
+        return values;
     }
 
-    /// @dev updates the last activity timestamp of a character
-    function _updateLastActivityTimestamp(uint256 characterID) private {
-        _lastActivityByCharacter[characterID] = block.timestamp;
+    /// @dev assigns an opponent to a character
+    function _assignOpponent(uint256 characterID) private {
+        uint8 tier = getArenaTier(characterID);
+        EnumerableSet.UintSet
+            storage matchableCharacters = _matchableCharactersByTier[tier];
+
+        require(matchableCharacters.length() != 0, "No enemy in tier");
+        require(!_duelQueue.contains(characterID), "Char dueling");
+
+        uint256 seed = randoms.getRandomSeed(msg.sender);
+        uint256 randomIndex = RandomUtil.randomSeededMinMax(
+            0,
+            matchableCharacters.length() - 1,
+            seed
+        );
+        uint256 opponentID;
+        uint256 matchableCharactersCount = matchableCharacters.length();
+        bool foundOpponent = false;
+
+        for (uint256 i = 0; i < matchableCharactersCount; i++) {
+            uint256 index = (randomIndex + 1) % matchableCharactersCount;
+            uint256 candidateID = matchableCharacters.at(index);
+
+            if (candidateID == characterID) {
+                if (matchableCharactersCount == 1) {
```

Comment:
This should be checked before the for loop rather than inside it.
Checking it in the loop is a waste of gas and the answer won't change while iterating.